File: musesp_ui/components/label.py
"""Label —— 文本标签组件。"""

import pygame

from musesp_ui.components.core import Component
from musesp_ui.font import get_font


class Label(Component):
    """文本标签。

    在组件区域内水平和竖直居中渲染文本。
    """

    def __init__(self, text: str = "", x: int = 0, y: int = 0, width: int = 0,
                 height: int = 0, font_size: int = 24,
                 color: tuple[int, int, int] = (255, 255, 255)):
        """:param text: 显示文本
        :param font_size: 字号
        :param color: RGB 颜色
        """
        super().__init__(x=x, y=y, width=width, height=height)
        self.text = text
        self.font_size = font_size
        self.color = color

    def _draw_self(self, surface: pygame.Surface, draw_x: int, draw_y: int) -> None:
        font = get_font(self.font_size)
        text_surface = font.render(self.text, True, self.color)
        tw, th = text_surface.get_size()
        cx = draw_x + (self.width - tw) // 2
        cy = draw_y + (self.height - th) // 2
        surface.blit(text_surface, (cx, cy))

File: musesp_ui/src/components/label.rs
use crate::components::core::{ComponentBase, ComponentTrait};
use crate::renderer::UIRenderer;

pub struct Label {
    pub base: ComponentBase,
    pub text: String,
    pub font_size: u32,
    pub color: (u8, u8, u8),
}

impl Label {
    pub fn new(text: &str, x: i32, y: i32, width: i32, height: i32, font_size: u32, color: (u8, u8, u8)) -> Box<Self> {
        Box::new(Label {
            base: ComponentBase::new(x, y, width, height),
            text: text.to_string(),
            font_size,
            color,
        })
    }
}

impl ComponentTrait for Label {
    fn base(&self) -> &ComponentBase { &self.base }
    fn base_mut(&mut self) -> &mut ComponentBase { &mut self.base }
    fn draw_self(&self, renderer: &mut UIRenderer, dx: i32, dy: i32) {
        renderer.draw_text(&self.text, dx, dy, self.base.width, self.base.height, self.font_size, self.color);
    }
}
